added keyboard controller

irwineffect opened the pull request:
useful for moving the sub around in the simulator

irwineffect committed:
added keyboard controller

skallaher commented on src/joystick/keyboard_control.cpp:
```diff
@@ -0,0 +1,81 @@
+
+uint8_t getKey(void)
+{
+    return fgetc(stdin);
```
This is using the C standard input. I recommend using cin and http://www.cplusplus.com/reference/istream/istream/get/ unless you require C style.

irwineffect replied:
Oh, the code was inspired by a few posts I found on stackoverflow lol.

skallaher commented on src/joystick/keyboard_control.cpp:
```diff
@@ -0,0 +1,81 @@
+    struct termios attr_backup;
+    tcgetattr(fileno(stdin), &attr_backup);
+
+    //setup new terminal settings
```
Information on what these are actually doing would be helpful.

skallaher commented on src/joystick/keyboard_control.cpp:
```diff
@@ -0,0 +1,81 @@
+                 break;
+            case 'c':
+                  msg.dive = 0.25;
+                  break;
```
Arrow key support? Perhaps a reset button like ESC?

bkallaher replied:
I've got arrow keys working. what should dive and yaw keys be?

bkallaher replied:
I'm thinking pg up and pg down for dive.

skallaher reviewed (changes requested):
Some things to think about.

bkallaher committed:
Added support for the arrow keys

Usage: Arrows move within the x/y plane (forward/strafe), PageUp/PageDn control dive, and pressing Ctrl with any of the arrow keys will do a rotation (left/right = yaw, up/down = pitch)

skallaher commented on src/joystick/keyboard_control.cpp:
```diff
@@ -41,34 +41,81 @@ int main(int argc, char **argv)
+                            msg.pitch_down = 10;
+                            break;
+                        case 'B':  // Down Key pressed
+                            msg.pitch_down = -10;
```
If the down key is meant to be pitch down, this should be positive. Positive `pitch_down` is pitching down. Also, pitch is absolute above. This would only allow for pitching up and down 10 degrees maximum.

bkallaher replied:
This is the way it is meant to be, with the up key pitching down. This emulates the way the joystick works.

skallaher replied:
Ah, it's inverted. Makes sense.

irwineffect committed:
renamed the joystick directory to be "controllers"

skallaher commented on src/controllers/keyboard_control.cpp:
```diff
@@ -0,0 +1,128 @@
+    while(ros::ok())
+    {
+        robosub::control msg = base_msg;
+        uint8_t key = getKey(), k1, k2;
```
Is there a reason these are uint8_ts rather than just chars?

irwineffect replied:
nah no reason

bkallaher committed:
Arrow keys fixed and working but better method is available

Will update with looping over key codes.

bkallaher committed:
Better key code parsing.

Modification can be done to allow movement in all axes if wanted

bkallaher committed:
Fixed Linter errors
